In [2]:
import numpy as np
import pandas as pd
from datetime import datetime
import matplotlib.pyplot as plt
import seaborn as sns
import missingno as msno

In [3]:
df = pd.read_csv("DataJobseeker.csv", encoding='ISO-8859-1')
df

C:\Users\lenovo\AppData\Local\Temp\ipykernel_3692\3364100549.py:1: DtypeWarning: Columns (4,27,29) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv("DataJobseeker.csv", encoding='ISO-8859-1')


js_unique_id request_date  daily_serial_no       eng_name_of_js  \
0           4848316   2016/03/01                1            RAJ KUMAR   
1           4848515   2015/05/01                1          Dilip Kumar   
2           4814815   01-48-2015                1             ABHIKESH   
3           4820216   2016/02/02                1         GULZAR AHMAD   
4           4820515   2015/05/02                1        RAHUL RATHOUR   
...             ...          ...              ...                  ...   
99995    1455180815   2015/08/18             1455  MAHENDRA VISHWKARMA   
99996    1455190216   2016/02/19             1455          ZIA TASNEEM   
99997    1455190615   2015/06/19             1455   SUDHIR KUMAR SUMAN   
99998    1455200316   2016/03/20             1455           SAID AHMAD   
99999    1455210116   2016/01/21             1455       PRASHANT KUMAR   

      phone_no_js                        js_email js_request_ip_address  \
0      8127302578        shiva.herbal26@gmail.com          59.97.88.199   
1      6307263126              dk680871@gmail.com        117.201.54.148   
2      8081643688     abhikeshverma4321@gmail.com        481.212.71.481   
3      7266044877         nadeemmau2016@gmail.com           1.39.48.144   
4      7607412349      eapulrathourboss@gmail.com         37.228.485.63   
...           ...                             ...                   ...   
99995  8853727261  mahendravishwakarma2@gmail.com       122.163.238.211   
99996  8423705426                    om@yahoo.com           1.39.51.243   
99997  9451211140      sudhir.1187@rediffmail.com          106.219.0.30   
99998  8869820800      sayeed.ahmad1985@gmail.com        112.110.23.143   
99999  9616485914     rajeshcomputergnj@gmail.com            1.39.48.60   

                      generated_1st_pwd pwd_sent_mode         pwd_sent_date  \
0      81dc9bdb52d04dc20036dbd8313ed055             B            2016/03/03   
1      c9d20dfa5c0dd98bc81dc145917edd87             B            2015/05/01   
2      df8ed725770e25ca30f27173a3b62aef             B  01-48-2015  00:01:05   
3      55017a6dfdca3a0cc433c9a2c96ef31c             B            2016/02/02   
4      a1ca5b6655ea0dba2dea6a6b5b5d8f14             B            2015/05/02   
...                                 ...           ...                   ...   
99995  b42136d96658ad0a8d9cec101dbc1423             B            2015/08/18   
99996  202ca662ac59075a664b07152d234b70             B            2016/02/19   
99997  4a0c0b3c4a9b0448806a2dccdebacf04             B            2015/06/19   
99998  e10adc3949ba59abbe56e057f20f883e             B            2016/03/20   
99999  79cfac6387ee1882f83a29a04d0bcdc4             B            2016/01/21   

       ... usernameEdistrict AadharFlag       Dob gender Aadhar_verify_dt  \
0      ...               NaN        NaN       NaN    NaN              NaN   
1      ...               NaN          Y  1/1/1996      M       2024/07/12   
2      ...               NaN        NaN       NaN    NaN              NaN   
3      ...               NaN        NaN       NaN    NaN              NaN   
4      ...               NaN        NaN       NaN    NaN              NaN   
...    ...               ...        ...       ...    ...              ...   
99995  ...               NaN        NaN       NaN    NaN              NaN   
99996  ...               NaN        NaN       NaN    NaN              NaN   
99997  ...               NaN        NaN       NaN    NaN              NaN   
99998  ...               NaN        NaN       NaN    NaN              NaN   
99999  ...               NaN        NaN       NaN    NaN              NaN   

      CheckFlag     userreg            insertdate  \
0             O     4848317            2016/03/01   
1             O     4848516            2015/05/01   
2             O     4814816  01-48-2015  00:00:00   
3             O     4820217            2016/02/02   
4             O     4820516            2015/05/02   
...         ...         .

| **Column**                               | **Description**                               |
| ---------------------------------------- | --------------------------------------------- |
| `js_unique_id`                           | Unique identifier for each job seeker         |
| `request_date`                           | Date of registration request                  |
| `daily_serial_no`                        | Daily sequence number                         |
| `eng_name_of_js`                         | Full name of the job seeker                   |
| `phone_no_js`, `js_email`                | Contact info                                  |
| `js_request_ip_address`                  | IP address used for registration              |
| `generated_1st_pwd`                      | Hashed password initially generated           |
| `pwd_sent_mode`, `pwd_sent_date`         | Mode/date of password delivery                |
| `js_first_log_date`                      | First login date                              |
| `LastloginDate`, `LoginAttempts`         | Login behavior tracking                       |
| `lock`, `NewPassword`, `NewPassFlag`     | Password status and login lock indicators     |
| `prf_submitted`, `final_submittion_date` | Whether profile was completed and when        |
| `satyapan_done`, `aft_satyapan_login_dt` | Aadhaar-based verification completed and when |
| `AadharFlag`, `Aadhar_verify_dt`         | Aadhaar verified or not, and date             |
| `Dob`, `gender`                          | Demographics                                  |
| `insertdate`                             | Date when this record was added to the system |


### Convert string columns to datetime format

#### Handle request_date column

In [4]:
# Valid dates will be converted properly to yyyy-mm-dd format
df['clean_request_date'] = pd.to_datetime(df['request_date'], errors = 'coerce')

In [ ]:
# Check how many rows were not converted
df['clean_request_date'].isna().sum()

In [ ]:
# See bad date strings
df[df['clean_request_date'].isna() & df['request_date'].notna()]['request_date'].value_counts()

# df['clean_request_date'].isna() - gave all the NaT rows from the 'clean_request_date' column
# df['request_date'].notna() - give all the rows that are not NULL in 'request_date'
# combining them ensures that we get only invalid values and not NULL values
# ['request_date'] - Pulls out just the original raw date values from those bad rows.

In [ ]:
# Understand the overall situation of the dataset i.e. how many columns of that row are null and what do they contain
df[df['clean_request_date'].isna()].describe(include='all')


In [ ]:
# Extract top wrong dates and undertand the common pattern
bad_dates = df[df['clean_request_date'].isna()]['request_date'].value_counts().head(15)
print(bad_dates)

SUMMARY- <br>
- around 850 rows have 48 in beginning 
- around 1100 rows have 48 as their middle value
- by handling these 2 problems we can make a large number of rows from invalid ones again useful 
- These are likely typos where 48 should’ve been a valid day
- **Strategy:** Replace '48-' with '04-' or '08-' based on context. You can assume the most common intended value 

In [9]:
# Step 1: Fix request_date values starting with '48' → replace '48' with '08'
mask_start_48 = df['request_date'].str.startswith('48', na=False)
df.loc[mask_start_48, 'clean_request_date'] = pd.to_datetime(
    df.loc[mask_start_48, 'request_date']
    .str.replace('--', '-', regex=False)
    .str.replace(r'^48', '08', regex=True),
    errors='coerce',
    dayfirst=True
)

# Step 2: Fix values where '48' is in the middle (but NOT those already handled in step 1)
mask_middle_48 = df['request_date'].str.contains('48', na=False) & ~mask_start_48
df.loc[mask_middle_48, 'clean_request_date'] = pd.to_datetime(
    df.loc[mask_middle_48, 'request_date']
    .str.replace('--', '-', regex=False)  # fix double hyphens
    .str.replace(r'-(48)-', '-08-', regex=True),
    errors='coerce',
    dayfirst=True
)

# Step 3: Fix values with '48' at both start and middle (but NOT those already handled in step 1 and 2)
mask_both_48 = df['request_date'].str.startswith('48', na=False) & df['request_date'].str.contains('48', na=False)
df.loc[mask_both_48, 'clean_request_date'] = pd.to_datetime(
    df.loc[mask_both_48, 'request_date']
    .str.replace('--', '-', regex=False)
    .str.replace(r'^48', '08', regex=True)
    .str.replace(r'-(48)-', '-08-', regex=True),
    errors='coerce',
    dayfirst=True
)

# Step-3 was written after executing the above 2 steps to understand how many rows are remaining and is there any pattern in them
# conclusion was that only those rows that have '48' at both start and middle were remaining so we handled them as well

In [10]:
df['clean_request_date'].isna().sum()

np.int64(0)

#### Handle pwd_sent_date column

In [11]:
# Valid dates will be converted properly to yyyy-mm-dd format
df['clean_pwd_sent_date'] = pd.to_datetime(df['pwd_sent_date'], errors='coerce')

In [12]:
# Check how many rows were not converted
df['clean_pwd_sent_date'].isna().sum()

np.int64(4121)

In [13]:
# See bad date strings
df[df['clean_pwd_sent_date'].isna() & df['pwd_sent_date'].notna()]['pwd_sent_date'].value_counts()

# df['clean_pwd_sent_date'].isna() - gave all the NaT rows from the 'clean_pwd_sent_date' column
# df['pwd_sent_date'].notna() - give all the rows that are not NULL in 'request_date'
# combining them ensures that we get only invalid values and not NULL values
# ['pwd_sent_date'] - Pulls out just the original raw date values from those bad rows.

pwd_sent_date
48-03-2016  48:22:01    2
29-48-2015  48:11:26    2
48-05-2015  48:29:00    2
48-03-2016  11:02:06    2
31-48-2015  07:09:00    2
                       ..
19-48-2015  09:52:59    1
20-48-2015  48:04:45    1
21-48-2015  48:13:49    1
25-48-2015  11:48:15    1
29-48-2015  11:17:57    1
Name: count, Length: 4098, dtype: int64

SUMMARY- <br>
- several rows have 48 in beginning and in middle
- by handling these 2 problems we can make a large number of rows from invalid ones again useful 
- These are likely typos where 48 should’ve been a valid day
- **Strategy:** Replace '48-' with '04-' or '08-' based on context. You can assume the most common intended value 

In [14]:
# Step 1: Fix pwd_sent_date values starting with '48' → replace '48' with '04', then remove time
mask_start_48 = df['pwd_sent_date'].str.startswith('48', na=False)
df.loc[mask_start_48, 'clean_pwd_sent_date'] = pd.to_datetime(
    df.loc[mask_start_48, 'pwd_sent_date']
    .str.replace(r'^48', '04', regex=True)       # replace starting 48 with 04
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],      # extract only the date part
    errors='coerce',
    dayfirst=True
)

# Step 2: Fix values where '48' is in the **middle** (not already handled in step 1), then remove time
mask_middle_48 = df['pwd_sent_date'].str.contains('48', na=False) & ~mask_start_48
df.loc[mask_middle_48, 'clean_pwd_sent_date'] = pd.to_datetime(
    df.loc[mask_middle_48, 'pwd_sent_date']
    .str.replace('--', '-', regex=False)  # fix double hyphens
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

# Step 3: Fix values with '48' at both start and middle (but NOT those already handled in step 1 and 2)
mask_both_48 = df['pwd_sent_date'].str.startswith('48', na=False) & df['pwd_sent_date'].str.contains('48', na=False)
df.loc[mask_both_48, 'clean_pwd_sent_date'] = pd.to_datetime(
    df.loc[mask_both_48, 'pwd_sent_date']
    .str.replace(r'^48', '08', regex=True)
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)


In [15]:
df['clean_pwd_sent_date'].isna().sum()

np.int64(0)

#### Handle js_first_log_date column

In [16]:
# Valid dates will be converted properly to yyyy-mm-dd format
df['clean_js_first_log_date'] = pd.to_datetime(df['js_first_log_date'], errors='coerce')

In [17]:
# Check how many rows were not converted
df['clean_js_first_log_date'].isna().sum()

np.int64(10055)

In [18]:
# See bad date strings
df[df['clean_js_first_log_date'].isna() & df['js_first_log_date'].notna()]['js_first_log_date'].value_counts()

# df['clean_js_first_log_date'].isna() - gave all the NaT rows from the 'clean_js_first_log_date' column
# df['js_first_log_date'].notna() - give all the rows that are not NULL in 'js_first_log_date'
# combining them ensures that we get only invalid values and not NULL values
# ['js_first_log_date'] - Pulls out just the original raw date values from those bad rows.

js_first_log_date
48-03-2016  11:09:36    2
03-48-2015  13:11:14    2
02-48-2015  15:01:38    2
02-48-2015  14:52:42    2
29-48-2015  09:44:03    2
                       ..
48-03-2016  00:05:21    1
48-12-2015  00:13:18    1
18-48-2015  00:19:41    1
01-48-2015  00:04:07    1
18-48-2015  00:05:50    1
Name: count, Length: 3831, dtype: int64

In [19]:
# Step 1: Fix pwd_sent_date values starting with '48' → replace '48' with '04', then remove time
mask_start_48 = df['js_first_log_date'].str.startswith('48', na=False)
df.loc[mask_start_48, 'clean_js_first_log_date'] = pd.to_datetime(
    df.loc[mask_start_48, 'js_first_log_date']
    .str.replace(r'^48', '04', regex=True)       # replace starting 48 with 04
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],      # extract only the date part
    errors='coerce',
    dayfirst=True
)

# Step 2: Fix values where '48' is in the **middle** (not already handled in step 1), then remove time
mask_middle_48 = df['js_first_log_date'].str.contains('48', na=False) & ~mask_start_48
df.loc[mask_middle_48, 'clean_js_first_log_date'] = pd.to_datetime(
    df.loc[mask_middle_48, 'js_first_log_date']
    .str.replace('--', '-', regex=False)  # fix double hyphens
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

# Step 3: Fix values with '48' at both start and middle (but NOT those already handled in step 1 and 2)
mask_both_48 = df['js_first_log_date'].str.startswith('48', na=False) & df['js_first_log_date'].str.contains('48', na=False)
df.loc[mask_both_48, 'clean_js_first_log_date'] = pd.to_datetime(
    df.loc[mask_both_48, 'js_first_log_date']
    .str.replace(r'^48', '08', regex=True)
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

C:\Users\lenovo\AppData\Local\Temp\ipykernel_3692\1857134237.py:3: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df.loc[mask_start_48, 'clean_js_first_log_date'] = pd.to_datetime(


In [20]:
df['clean_js_first_log_date'].isna().sum()

np.int64(6204)

In [21]:
# See bad date strings
df[df['clean_js_first_log_date'].isna() & df['js_first_log_date'].notna()]['js_first_log_date'].value_counts()

Series([], Name: count, dtype: int64)

Since the number of NULL is still shown 6204 but the above code to get non-null invalid values is showing nothing<br>
**This means that the remaining 6204 rows are NULL**

#### Handle final_submittion_date column

In [22]:
# Valid dates will be converted properly to yyyy-mm-dd format
df['clean_final_submittion_date'] = pd.to_datetime(df['final_submittion_date'], errors='coerce')

In [23]:
# Check how many rows were not converted
df['clean_final_submittion_date'].isna().sum()

np.int64(7638)

In [24]:
# See bad date strings
df[df['clean_final_submittion_date'].isna() & df['final_submittion_date'].notna()]['final_submittion_date'].value_counts()

# df['clean_final_submittion_date'].isna() - gave all the NaT rows from the 'clean_final_submittion_date' column
# df['final_submittion_date'].notna() - give all the rows that are not NULL in 'final_submittion_date'
# combining them ensures that we get only invalid values and not NULL values
# ['final_submittion_date'] - Pulls out just the original raw date values from those bad rows.

final_submittion_date
48-02-2016  14:59:36    2
48-02-2016  11:00:45    2
48-03-2016  12:12:26    2
48-03-2016  48:56:30    2
14-48-2015  06:44:13    2
                       ..
07-48-2015  14:12:28    1
19-48-2015  48:06:23    1
20-48-2015  48:20:04    1
21-48-2015  48:20:51    1
09-48-2015  00:48:44    1
Name: count, Length: 4119, dtype: int64

In [25]:
# Step 1: Fix final_submittion_date values starting with '48' → replace '48' with '04', then remove time
mask_start_48 = df['final_submittion_date'].str.startswith('48', na=False)
df.loc[mask_start_48, 'clean_final_submittion_date'] = pd.to_datetime(
    df.loc[mask_start_48, 'final_submittion_date']
    .str.replace(r'^48', '04', regex=True)       # replace starting 48 with 04
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],      # extract only the date part
    errors='coerce',
    dayfirst=True
)

# Step 2: Fix values where '48' is in the **middle** (not already handled in step 1), then remove time
mask_middle_48 = df['final_submittion_date'].str.contains('48', na=False) & ~mask_start_48
df.loc[mask_middle_48, 'clean_final_submittion_date'] = pd.to_datetime(
    df.loc[mask_middle_48, 'final_submittion_date']
    .str.replace('--', '-', regex=False)  # fix double hyphens
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

# Step 3: Fix values with '48' at both start and middle (but NOT those already handled in step 1 and 2)
mask_both_48 = df['final_submittion_date'].str.startswith('48', na=False) & df['final_submittion_date'].str.contains('48', na=False)
df.loc[mask_both_48, 'clean_final_submittion_date'] = pd.to_datetime(
    df.loc[mask_both_48, 'final_submittion_date']
    .str.replace(r'^48', '08', regex=True)
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

In [26]:
df['clean_final_submittion_date'].isna().sum()

np.int64(3510)

In [27]:
df[df['clean_final_submittion_date'].isna() & df['final_submittion_date'].notna()]['final_submittion_date'].value_counts()

Series([], Name: count, dtype: int64)

Since the number of NULL is still shown 3510 but the above code to get non-null invalid values is showing nothing<br>
**This means that the remaining 3510 rows are NULL**

#### Handle insertdate column

In [28]:
# Valid dates will be converted properly to yyyy-mm-dd format
df['clean_insertdate'] = pd.to_datetime(df['insertdate'], errors='coerce')

In [29]:
# Check how many rows were not converted
df['clean_insertdate'].isna().sum()

np.int64(4121)

In [30]:
# See bad date strings
df[df['clean_insertdate'].isna() & df['insertdate'].notna()]['insertdate'].value_counts()

# df['clean_insertdate'].isna() - gave all the NaT rows from the 'clean_insertdate' column
# df['insertdate'].notna() - give all the rows that are not NULL in 'insertdate'
# combining them ensures that we get only invalid values and not NULL values
# ['insertdate'] - Pulls out just the original raw date values from those bad rows.

insertdate
48-03-2016  00:00:00    338
48-02-2016  00:00:00    303
48-48-2015  00:00:00    130
13-48-2015  00:00:00    126
48-05-2015  00:00:00    120
04-48-2015  00:00:00    113
12-48-2015  00:00:00    112
11-48-2015  00:00:00    108
23-48-2015  00:00:00    106
26-48-2015  00:00:00    105
14-48-2015  00:00:00    104
21-48-2015  00:00:00    102
16-48-2015  00:00:00    101
20-48-2015  00:00:00    100
18-48-2015  00:00:00    100
08-48-2015  00:00:00     99
19-48-2015  00:00:00     99
25-48-2015  00:00:00     98
01-48-2015  00:00:00     98
22-48-2015  00:00:00     97
06-48-2015  00:00:00     96
27-48-2015  00:00:00     96
07-48-2015  00:00:00     93
09-48-2015  00:00:00     93
31-48-2015  00:00:00     90
15-48-2015  00:00:00     90
28-48-2015  00:00:00     87
24-48-2015  00:00:00     84
48-12-2015  00:00:00     81
29-48-2015  00:00:00     80
48-06-2015  00:00:00     78
17-48-2015  00:00:00     76
02-48-2015  00:00:00     75
03-48-2015  00:00:00     74
30-48-2015  00:00:00     74
05-48-201

In [31]:
# Step 1: Fix insertdate values starting with '48' → replace '48' with '04', then remove time
mask_start_48 = df['insertdate'].str.startswith('48', na=False)
df.loc[mask_start_48, 'clean_insertdate'] = pd.to_datetime(
    df.loc[mask_start_48, 'insertdate']
    .str.replace(r'^48', '04', regex=True)       # replace starting 48 with 04
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],      # extract only the date part
    errors='coerce',
    dayfirst=True
)

# Step 2: Fix values where '48' is in the **middle** (not already handled in step 1), then remove time
mask_middle_48 = df['insertdate'].str.contains('48', na=False) & ~mask_start_48
df.loc[mask_middle_48, 'clean_insertdate'] = pd.to_datetime(
    df.loc[mask_middle_48, 'insertdate']
    .str.replace('--', '-', regex=False)  # fix double hyphens
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

# Step 3: Fix values with '48' at both start and middle (but NOT those already handled in step 1 and 2)
mask_both_48 = df['insertdate'].str.startswith('48', na=False) & df['insertdate'].str.contains('48', na=False)
df.loc[mask_both_48, 'clean_insertdate'] = pd.to_datetime(
    df.loc[mask_both_48, 'insertdate']
    .str.replace(r'^48', '08', regex=True)
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

In [32]:
df['clean_insertdate'].isna().sum()

np.int64(0)

#### Handle Dob column

In [33]:
# Valid dates will be converted properly to yyyy-mm-dd format
df['clean_Dob'] = pd.to_datetime(df['Dob'], errors='coerce')

In [34]:
# Check how many rows were not converted
df['clean_Dob'].isna().sum()

np.int64(94063)

In [35]:
# See bad date strings
df[df['clean_Dob'].isna() & df['Dob'].notna()]['Dob'].value_counts()

# df['clean_Dob'].isna() - gave all the NaT rows from the 'clean_Dob' column
# df['Dob'].notna() - give all the rows that are not NULL in 'Dob'
# combining them ensures that we get only invalid values and not NULL values
# ['Dob'] - Pulls out just the original raw date values from those bad rows.

Dob
48-07-1992    10
48-07-1994     9
48-07-1997     8
48-07-1990     8
48-08-1994     7
              ..
07-48-1992     1
48-12-1991     1
48-06-1989     1
48-06-1980     1
48-09-1987     1
Name: count, Length: 257, dtype: int64

In [36]:
# Step 1: Fix Dob values starting with '48' → replace '48' with '04', then remove time
mask_start_48 = df['Dob'].str.startswith('48', na=False)
df.loc[mask_start_48, 'Dob'] = pd.to_datetime(
    df.loc[mask_start_48, 'Dob']
    .str.replace(r'^48', '04', regex=True),       # replace starting 48 with 04
    errors='coerce',
    dayfirst=True
)

# Step 2: Fix values where '48' is in the **middle** (not already handled in step 1), then remove time
mask_middle_48 = df['Dob'].str.contains('48', na=False) & ~mask_start_48
df.loc[mask_middle_48, 'clean_Dob'] = pd.to_datetime(
    df.loc[mask_middle_48, 'Dob']
    .str.replace('--', '-', regex=False)  # fix double hyphens
    .str.replace(r'-(48)-', '-08-', regex=True),
    errors='coerce',
    dayfirst=True
)

# Step 3: Fix values with '48' at both start and middle (but NOT those already handled in step 1 and 2)
mask_both_48 = df['Dob'].str.startswith('48', na=False) & df['Dob'].str.contains('48', na=False)
df.loc[mask_both_48, 'clean_Dob'] = pd.to_datetime(
    df.loc[mask_both_48, 'Dob']
    .str.replace(r'^48', '08', regex=True)
    .str.replace(r'-(48)-', '-08-', regex=True),
    errors='coerce',
    dayfirst=True
)

In [37]:
df['clean_Dob'].isna().sum()

np.int64(93905)

In [38]:
df[df['clean_final_submittion_date'].isna() & df['final_submittion_date'].notna()]['final_submittion_date'].value_counts()

Series([], Name: count, dtype: int64)

Since the number of NULL is still shown 93606 but the above code to get non-null invalid values is showing nothing<br>
**This means that the remaining 93606 rows are NULL**

#### Handle LastloginDate columns

In [39]:
# Valid dates will be converted properly to yyyy-mm-dd format
df['clean_LastloginDate'] = pd.to_datetime(df['LastloginDate'], errors='coerce')

In [40]:
# Check how many rows were not converted
df['clean_LastloginDate'].isna().sum()

np.int64(55127)

In [41]:
# See bad date strings
df[df['clean_LastloginDate'].isna() & df['LastloginDate'].notna()]['LastloginDate'].value_counts()

# df['clean_LastloginDate'].isna() - gave all the NaT rows from the 'clean_LastloginDate' column
# df['LastloginDate'].notna() - give all the rows that are not NULL in 'LastloginDate'
# combining them ensures that we get only invalid values and not NULL values
# ['LastloginDate'] - Pulls out just the original raw date values from those bad rows.

LastloginDate
22-48-2021  16:14:24    1
13-48-2021  22:34:18    1
48-08-2022  23:37:46    1
01-48-2022  17:24:18    1
02-48-2021  20:57:43    1
                       ..
48-04-2018  17:32:47    1
21-48-2016  18:23:32    1
21-48-2021  14:51:38    1
48-04-2025  11:51:05    1
24-48-2021  23:06:24    1
Name: count, Length: 1873, dtype: int64

In [42]:
# Step 1: Fix LastloginDate values starting with '48' → replace '48' with '04', then remove time
mask_start_48 = df['LastloginDate'].str.startswith('48', na=False)
df.loc[mask_start_48, 'clean_LastloginDate'] = pd.to_datetime(
    df.loc[mask_start_48, 'LastloginDate']
    .str.replace(r'^48', '04', regex=True)       # replace starting 48 with 04
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],      # extract only the date part
    errors='coerce',
    dayfirst=True
)

# Step 2: Fix values where '48' is in the **middle** (not already handled in step 1), then remove time
mask_middle_48 = df['LastloginDate'].str.contains('48', na=False) & ~mask_start_48
df.loc[mask_middle_48, 'clean_LastloginDate'] = pd.to_datetime(
    df.loc[mask_middle_48, 'LastloginDate']
    .str.replace('--', '-', regex=False)  # fix double hyphens
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

# Step 3: Fix values with '48' at both start and middle (but NOT those already handled in step 1 and 2)
mask_both_48 = df['LastloginDate'].str.startswith('48', na=False) & df['LastloginDate'].str.contains('48', na=False)
df.loc[mask_both_48, 'clean_LastloginDate'] = pd.to_datetime(
    df.loc[mask_both_48, 'LastloginDate']
    .str.replace(r'^48', '08', regex=True)
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

In [43]:
df['clean_LastloginDate'].isna().sum()

np.int64(53254)

In [44]:
df[df['clean_LastloginDate'].isna() & df['LastloginDate'].notna()]['LastloginDate'].value_counts()

Series([], Name: count, dtype: int64)

Since the number of NULL is still shown 53254 but the above code to get non-null invalid values is showing nothing<br>
**This means that the remaining 53254 rows are NULL**

#### Handle Aadhar_verify_dt column

In [45]:
# Valid dates will be converted properly to yyyy-mm-dd format
df['clean_Aadhar_verify_dt'] = pd.to_datetime(df['Aadhar_verify_dt'], errors='coerce')

In [46]:
# Check how many rows were not converted
df['clean_Aadhar_verify_dt'].isna().sum()

np.int64(93929)

In [47]:
# See bad date strings
df[df['clean_Aadhar_verify_dt'].isna() & df['Aadhar_verify_dt'].notna()]['Aadhar_verify_dt'].value_counts()

# df['clean_Aadhar_verify_dt'].isna() - gave all the NaT rows from the 'clean_Aadhar_verify_dt' column
# df['Aadhar_verify_dt'].notna() - give all the rows that are not NULL in 'LastloginDate'
# combining them ensures that we get only invalid values and not NULL values
# ['Aadhar_verify_dt'] - Pulls out just the original raw date values from those bad rows.

Aadhar_verify_dt
48-03-2023  13:33:34    1
12-48-2023  11:09:57    1
26-48-2023  14:39:50    1
48-09-2024  08:02:39    1
48-03-2023  48:17:53    1
                       ..
01-48-2023  19:54:38    1
11-48-2023  17:52:08    1
48-07-2023  13:32:28    1
26-48-2023  18:57:30    1
48-06-2023  16:55:52    1
Name: count, Length: 323, dtype: int64

In [48]:
# Step 1: Fix Aadhar_verify_dt values starting with '48' → replace '48' with '04', then remove time
mask_start_48 = df['Aadhar_verify_dt'].str.startswith('48', na=False)
df.loc[mask_start_48, 'clean_Aadhar_verify_dt'] = pd.to_datetime(
    df.loc[mask_start_48, 'Aadhar_verify_dt']
    .str.replace(r'^48', '04', regex=True)       # replace starting 48 with 04
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],      # extract only the date part
    errors='coerce',
    dayfirst=True
)

# Step 2: Fix values where '48' is in the **middle** (not already handled in step 1), then remove time
mask_middle_48 = df['Aadhar_verify_dt'].str.contains('48', na=False) & ~mask_start_48
df.loc[mask_middle_48, 'clean_Aadhar_verify_dt'] = pd.to_datetime(
    df.loc[mask_middle_48, 'Aadhar_verify_dt']
    .str.replace('--', '-', regex=False)  # fix double hyphens
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

# Step 3: Fix values with '48' at both start and middle (but NOT those already handled in step 1 and 2)
mask_both_48 = df['Aadhar_verify_dt'].str.startswith('48', na=False) & df['Aadhar_verify_dt'].str.contains('48', na=False)
df.loc[mask_both_48, 'clean_Aadhar_verify_dt'] = pd.to_datetime(
    df.loc[mask_both_48, 'Aadhar_verify_dt']
    .str.replace(r'^48', '08', regex=True)
    .str.replace(r'-(48)-', '-08-', regex=True)
    .str.extract(r'(\d{2}-\d{2}-\d{4})')[0],
    errors='coerce',
    dayfirst=True
)

In [49]:
df['clean_Aadhar_verify_dt'].isna().sum()

np.int64(93606)

In [50]:
df[df['clean_Aadhar_verify_dt'].isna() & df['Aadhar_verify_dt'].notna()]['Aadhar_verify_dt'].value_counts()

Series([], Name: count, dtype: int64)

Since the number of NULL is still shown 93606 but the above code to get non-null invalid values is showing nothing<br>
**This means that the remaining 93606 rows are NULL**

### Format Categorical Fields

In [51]:
df['pwd_sent_mode'].value_counts(dropna=False)

pwd_sent_mode
B    89399
M    10601
Name: count, dtype: int64

- B -> Batch i.e. generated by auto-suggest medium
- M -> Manual i.e. user-generated by specific individual

In [52]:
df['prf_submitted'].value_counts(dropna=False)

prf_submitted
Y      44261
NaN    29566
D      26173
Name: count, dtype: int64

- Y -> Yes i.e. complete profile created and submitted
- D -> Draft i.e. profile is saved but not submitted
- NaN -> Null i.e. profile not created

In [53]:
df['lock'].value_counts(dropna=False)

lock
0    100000
Name: count, dtype: int64

In [54]:
df['LoginAttempts'].value_counts(dropna=False)

LoginAttempts
0     91480
1      4410
2      1851
3       812
4       680
5       332
6       191
7       121
8        64
9        58
48        1
Name: count, dtype: int64

In [55]:
df['userActiveDeactive'].value_counts(dropna=False)

userActiveDeactive
NaN    84845
1.0     7869
0.0     7286
Name: count, dtype: int64

In [56]:
df['userActiveDeactiveToken'].value_counts(dropna=False)

userActiveDeactiveToken
NaN                                 92131
54a367d629152b720749e187b3eaa11b       20
07563a3fe3bbe7e3ba84431ad9d055af       17
a3d68b461bd9d3533ee1dd3ce4628ed4       16
432aca3a1e345e339f35a30c8f65edce       16
                                    ...  
0bb4aec1748521c12ee76289d9440817        1
2f37d10131f2a483a8dd005b3d14b0d9        1
cd00692c3bfe59267d5ecfac5348286c        1
4b6538a44a1dfdc2b83477cd76dee98e        1
b6a1085a27ab7bff7550f8a3bd017df8        1
Name: count, Length: 1093, dtype: int64

#### Create New Columns (Feature Engineering)

In [ ]:
today_date = pd.to_datetime('today')  # includes time

df['account_age_days'] = (today_date - df['clean_insertdate']).dt.days # get only the number of days
df

In [ ]:
df['is_verified'] = df['AadharFlag'].apply(lambda x: 1 if x == 'Y' else 0)
df

In [ ]:
df['is_active_user'] = df['LoginAttempts'].apply(lambda x: 1 if x > 0 else 0)
df

In [ ]:
df['has_logged_in'] = df['js_first_log_date'].apply(lambda x: 1 if pd.notna(x) else 0) 
df

### Checking and Handling Duplicates

In [61]:
df['js_unique_id'].duplicated().sum()


np.int64(449)

In [ ]:
# Get the unique ids which appeared more than once
df['js_unique_id'].value_counts()[df['js_unique_id'].value_counts() > 1]


In [63]:
# Check if the same id is allocated to the same person
df.groupby(['js_unique_id', 'eng_name_of_js']).size().reset_index(name='count').query('count > 1')


Empty DataFrame
Columns: [js_unique_id, eng_name_of_js, count]
Index: []

#### Summary of Duplicate Handling
There are **449** such **'js_unique_id'** which are given to 2 completely different persons and it can lead to serious issues if not handled

### Final Summary after preprocessing

| Column                                                                   | Reason to leave as it is & not use in analysis   |
| ------------------------------------------------------------------------ | ------------------------------------------------ |
| `usernameEdistrict`                                                      | 100% null                                        |
| `Is_Archival`, `ArogyaSetu`, `satyapan_done`, `CheckFlag`, `AadharFlag`, `old_reg_YN`, `lock` | Single value only           |
| `NewPassword`, `NewPassFlag`, `userActiveDeactiveToken`, `generated_1st_pwd`, `current_pwd`| Sensitive or hashed passwords  |
| `aft_satyapan_login_dt`, `sw_unique_id_create_date`                      | Data in wrong format expected date but is time   |
| `Dob`, `gender`, `Aadhar_verify_dt`, `csc_request_key`, `csc_user_type`  | More than 93000 rows are missing                 |
| `daily_serial_no`                                                        | Likely sequential and not useful                 |
| `username`, `sw_unique_id`                                               | Not sure what it shows                           |
| `question_id`, `security_answer`, `userActiveDeactive`             | Not relevant unless you're analyzing security patterns |

### Renaming the columns

In [ ]:
df = df.rename(columns={
    'eng_name_of_js': 'js_name',
    'phone_no_js': 'js_phoneNo', 
    'js_browser': 'js_browser_version',
    'js_browser_ver' : 'js_browser_name'
})

df

#### Saving as a cleaned and preprocessed file

In [65]:
columns_to_save = ['js_unique_id', 'js_name', 'js_phoneNo', 'js_email', 'js_request_ip_address', 'pwd_sent_mode', 'js_browser_name', 'js_browser_version', 'prf_submitted', 'satyapan_done', 'LoginAttempts', 'clean_request_date', 'clean_pwd_sent_date', 'clean_js_first_log_date', 'clean_final_submittion_date', 'clean_insertdate',
       'clean_LastloginDate', 'clean_Aadhar_verify_dt', 'account_age_days',
       'is_verified', 'is_active_user', 'has_logged_in']

df[columns_to_save].to_csv('Preprocessed_DataJobSeeker.csv', index=False)